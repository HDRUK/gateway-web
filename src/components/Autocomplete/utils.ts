import { ValueType } from "./Autocomplete";

const getChipLabel = (
    options: { value: string | number; label: string }[],
    value: ValueType
) => {
    if (typeof value === "string") {
        return value;
    }

    return options.find(option => option.value === value)?.label;
};

export { getChipLabel };
