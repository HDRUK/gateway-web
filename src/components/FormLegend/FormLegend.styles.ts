import { styled } from "@mui/material";
import Box from "@/components/Box";
import Typography from "@/components/Typography";

interface WrapperProps {
    offsetTop: string;
}

interface ProgressIconProps {
    iconColour: string;
}

const ICON_SIZE = "18px";
const SVG_SIZE = "13px";

export const Wrapper = styled(Box, {
    shouldForwardProp: prop => prop !== "offsetTop",
})<WrapperProps>(({ offsetTop }) => ({
    position: "sticky",
    top: offsetTop,
    padding: 0,
}));

export const LegendIcon = styled("span")<ProgressIconProps>(
    ({ iconColour }) => ({
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        width: ICON_SIZE,
        height: ICON_SIZE,
        backgroundColor: iconColour,
        borderRadius: "50%",
        marginRight: "10px",
        flexShrink: 0,

        "& svg": {
            color: "white",
            fontSize: SVG_SIZE,
        },
    })
);

export const BookmarkText = styled(Typography)(({ theme }) => ({
    padding: `${theme.spacing(3)} ${theme.spacing(1)}`,
    borderBottom: `1px solid ${theme.palette.greyCustom.main}`,
    margin: `0 ${theme.spacing(2)}`,
}));
