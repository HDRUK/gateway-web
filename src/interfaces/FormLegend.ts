export enum LegendStatus {
    VALID,
    OPTIONAL_REMAIN,
    INVALID,
    ACTIVE,
    UNTOUCHED,
}

export type LegendItem = {
    name: string;
    status: LegendStatus;
};
