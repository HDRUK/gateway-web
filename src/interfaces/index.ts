import { Error, ErrorItem } from "./Error";
import { User } from "./User";

export type { Error, ErrorItem, User };
