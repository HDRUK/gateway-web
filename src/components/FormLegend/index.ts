import FormLegend from "./FormLegend";

export default FormLegend;
