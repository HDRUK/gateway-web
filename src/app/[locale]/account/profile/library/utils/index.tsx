import { IconButton } from "@mui/material";
import { createColumnHelper } from "@tanstack/react-table";
import { LibraryListItem, SelectedLibrary } from "@/interfaces/Library";
import Link from "@/components/Link";
import StyledCheckbox from "@/components/StyledCheckbox";
import TooltipIcon from "@/components/TooltipIcon";
import { CheckIcon, DeleteForeverIcon } from "@/consts/icons";
import { RouteName } from "@/consts/routeName";

const columnHelper = createColumnHelper<LibraryListItem>();

const getColumns = ({
    handleSelect,
    handleRemove,
    selected,
    translations,
}: {
    handleSelect: (data: SelectedLibrary) => void;
    handleRemove: (id: string | number) => void;
    selected: SelectedLibrary;
    translations: { [id: string]: string };
}) => [
    columnHelper.display({
        id: "actions",
        meta: { isPinned: true },
        cell: ({ row }) => {
            const {
                id,
                datasetId,
                name,
                dataCustodianId,
                dataCustodian,
                darEnabled,
                darTemplatePublished,
                cohortEnabled,
            } = row.original;
            return (
                <div style={{ textAlign: "center" }}>
                    <StyledCheckbox
                        checked={selected[id]?.selected}
                        onChange={(_e, value) =>
                            handleSelect({
                                [id]: {
                                    selected: value,
                                    datasetId,
                                    name,
                                    teamId: dataCustodianId,
                                    teamName: dataCustodian,
                                    darEnabled,
                                    darTemplatePublished,
                                    cohortEnabled,
                                },
                            })
                        }
                        size="large"
                        sx={{ p: 0 }}
                        iconSx={{ mr: 0 }}
                    />
                </div>
            );
        },
        header: () => null,
        size: 43,
    }),

    columnHelper.display({
        id: "name",
        cell: ({ row: { original } }) => {
            const { datasetId, name } = original;
            const linkHref = `/${RouteName.DATASET_ITEM}/${datasetId}`;

            return (
                <Link href={linkHref}>
                    <div style={{ textAlign: "center" }}>{name}</div>
                </Link>
            );
        },
        header: () => <span>{translations.name}</span>,
        size: 150,
    }),

    columnHelper.display({
        id: "darEnabled",
        cell: ({
            row: {
                original: { darEnabled },
            },
        }) => (
            <div style={{ textAlign: "center" }}>
                {darEnabled ? <CheckIcon color="primary" /> : "-"}
            </div>
        ),
        header: () => (
            <TooltipIcon
                buttonSx={{ p: 0 }}
                size="small"
                label={translations.darEnabled}
                content={translations.darEnabled}
            />
        ),
        size: 100,
    }),

    columnHelper.display({
        id: "cohortEnabled",
        cell: ({
            row: {
                original: { cohortEnabled },
            },
        }) => (
            <div style={{ textAlign: "center" }}>
                {cohortEnabled ? <CheckIcon color="primary" /> : "-"}
            </div>
        ),
        header: () => <span>{translations.cohortEnabled}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "dataCustodian",
        cell: ({
            row: {
                original: { dataCustodian },
            },
        }) => <div style={{ textAlign: "center" }}>{dataCustodian}</div>,
        header: () => <span>{translations.dataCustodian}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "entityType",
        cell: ({
            row: {
                original: { entityType },
            },
        }) => <div style={{ textAlign: "center" }}>{entityType}</div>,
        header: () => <span>{translations.entityType}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "delete",
        cell: ({ row: { original } }) => {
            return (
                <div style={{ textAlign: "center" }}>
                    <IconButton onClick={() => handleRemove(original.id)}>
                        <DeleteForeverIcon color="primary" />
                    </IconButton>
                </div>
            );
        },
        header: () => null,
        size: 43,
    }),
];

export { getColumns };
