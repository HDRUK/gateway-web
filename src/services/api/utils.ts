import { NotificationOptions } from "@/interfaces/Api";
import { AxiosResponse } from "axios";
import { Error } from "@/interfaces/Error";
import notificationService from "../notification";

interface ErrorNotificationProps {
    props: NotificationOptions;
    errorResponse: AxiosResponse<Error>;
    method: "delete" | "post" | "put" | "get";
}

const errorNotification = ({
    errorResponse,
    method,
    props,
}: ErrorNotificationProps) => {
    const { t, i18n, ...notificationProps } = props;
    const { data, status } = errorResponse || {};

    const fallbackTitle = i18n.exists(`api:common.error.status.${status}`)
        ? t(`api:common.error.status.${status}`)
        : "There has been an error";

    const title = data?.title || fallbackTitle;
    const message =
        data?.message ||
        t(`api:common.error.${method}.message`, {
            item: props.itemName || "Item",
        });

    notificationService.apiError(message, {
        title,
        message,
        errors: data?.errors,
        ...notificationProps,
    });
};

interface SuccessNotificationProps {
    props: NotificationOptions;
    method: "delete" | "post" | "put";
}

const successNotification = ({ props, method }: SuccessNotificationProps) => {
    const { t, i18n, ...notificationProps } = props;
    const customMessage = `api:${props.localeKey}.success.${method}.message`;
    const shouldOverideMessage = i18n.exists(customMessage);

    const message = shouldOverideMessage
        ? t(customMessage)
        : t(`api:common.success.${method}.message`, {
              item: props.itemName || "Item",
          });

    notificationService.success(message, { ...notificationProps });
};

export { errorNotification, successNotification };
