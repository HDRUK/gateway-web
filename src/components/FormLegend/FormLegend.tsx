"use client";

import { ListItemButton } from "@mui/material";
import { LegendItem, LegendStatus } from "@/interfaces/FormLegend";
import { colors } from "@/config/theme";
import { CloseIcon, CheckIcon, PriorityHighIcon } from "@/consts/icons";
import { capitalise, splitCamelcase } from "@/utils/general";
import Typography from "../Typography";
import { Wrapper, LegendIcon } from "./FormLegend.styles";

interface FormLegendProps {
    items: LegendItem[];
    offsetTop?: string;
    handleClickItem?: (itemIndex: number) => void;
}

const getIcon = (status: LegendStatus) => {
    switch (status) {
        case LegendStatus.VALID:
            return <CheckIcon />;
        case LegendStatus.OPTIONAL_REMAIN:
            return <PriorityHighIcon />;
        case LegendStatus.INVALID:
            return <CloseIcon />;
        default:
            return null;
    }
};

const getBackgroundColour = (status: LegendStatus) => {
    switch (status) {
        case LegendStatus.VALID:
            return colors.green400;
        case LegendStatus.ACTIVE:
            return colors.purple500;
        case LegendStatus.OPTIONAL_REMAIN:
            return "#F0BB24";
        case LegendStatus.INVALID:
            return colors.red700;
        default:
            return colors.purple100;
    }
};

const FormLegend = ({ items, offsetTop, handleClickItem }: FormLegendProps) => (
    <Wrapper offsetTop={offsetTop || "initial"}>
        {items.map((item, index) => (
            <ListItemButton
                key={item.name}
                onClick={() => handleClickItem && handleClickItem(index)}>
                <LegendIcon iconColour={getBackgroundColour(item.status)}>
                    {getIcon(item.status)}
                </LegendIcon>
                <Typography>{capitalise(splitCamelcase(item.name))}</Typography>
            </ListItemButton>
        ))}
    </Wrapper>
);

export default FormLegend;
