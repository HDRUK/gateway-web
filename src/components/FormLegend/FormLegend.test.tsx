import { LegendItem, LegendStatus } from "@/interfaces/FormLegend";
import { render } from "@/utils/testUtils";
import FormLegend from "./FormLegend";

describe("FormLegend", () => {
    const data: LegendItem[] = [{ name: "ITEM1", status: LegendStatus.VALID }];

    it("should render legend", async () => {
        const wrapper = render(<FormLegend items={data} />);
        expect(wrapper.container).toMatchSnapshot();
    });
});
