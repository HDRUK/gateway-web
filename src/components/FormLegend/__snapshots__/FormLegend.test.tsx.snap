// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormLegend should render legend 1`] = `
<div>
  <div
    class="MuiBox-root mui-style-171zbuw"
    offsettop="initial"
  >
    <div
      class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters mui-style-1hhp30x-MuiButtonBase-root-MuiListItemButton-root"
      role="button"
      tabindex="0"
    >
      <span
        class="mui-style-8tkev2"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-style-i4bv87-MuiSvgIcon-root"
          data-testid="CheckIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M9 16.17 4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
          />
        </svg>
      </span>
      <p
        class="MuiTypography-root MuiTypography-body1 mui-style-1t8ylhr-MuiTypography-root"
      >
        Item1
      </p>
      <span
        class="MuiTouchRipple-root mui-style-8je8zh-MuiTouchRipple-root"
      />
    </div>
  </div>
</div>
`;
